### DISCUSSION TITLE
Question about ART's Black Box Attack on Tesseract notebook example

### DISCUSSION THREAD

**Discussion by HexHexHex16** [Trusted-AI/adversarial-robustness-toolbox#1519] [Q&A]
**ART Black Box Attack on Tesseract example:**
https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/fd9e54892386a32b32e10e87fe86bfa22be1e3e8/notebooks/classifier_blackbox_tesseract.ipynb

As far as I understand notebook should have worked like this:
1) Import libraries
2) Reading picture 1 (used for overlay) and picture 2 (on which is superimposed)
3) Setting the prediction function to call tesseract from the command line and transform
4) Creating a dictionary with the keys of recognizable words on  pictures  1 and 2
5) Word recognition in  picture 1
6) Word recognition in  picture 2
7) Attack HopSkipJump  (overlay  picture 1 on 2)
 _{Result: Recognition of the word "dissent" in the processed picture}_
8) Apply JPEG-COMPRESSION
9) Word recognition in picture 1
10) Word recognition in picture 2
11) Attack HopSkipJump  (overlay  picture 1 on 2)
_{Result: Recognition of the word "assent" in the processed picture}_

But, the problem is that in 7) the processed "assent image " is recognized as "assent" and not as "dissent" during the attack, although the processed image looks like "dissent".
 To work correctly, the result on 7) must be "dissent" after the attack, and on 11) "assent", becasue in 8) the JPEG-COMPRESSION preprocessing method will be used

**Comment by beat-buesser** [comment]
Hi @HexHexHex16 I think you are right, cell 7 should run for more iterations until Tesseract predicts dissent. Cell 11 will also eventually, with more iterations show dissent, but the notebook aimed to show that a larger perturbation is required.

Let me know if you are updating the notebook, I'd be happy to pull it into the main branch.

**Reply by HexHexHex16** [reply]
Hi @beat-buesser 
I changed the number of iterations in cell 7) in the for loop from 16 to 32:

```
attack = HopSkipJump(classifier=classifier, targeted=True, norm=2, max_iter=0, max_eval=1000, init_eval=10)
iter_step = 10
x_adv = np.array([image_init], dtype=np.float32)
for i in range(32):
    x_adv = attack.generate(x=np.array([image_target], dtype=np.float32), x_adv_init=x_adv, y=to_categorical([1], 3))

    #clear_output()
    if i%3 == 0:
        print("Adversarial image at step %d." % (i * iter_step), "L2 error", 
              np.linalg.norm(np.reshape(x_adv[0] - image_target, [-1])),
              "and Tesseract output %s." % label_dict[np.argmax(classifier.predict(x_adv)[0])])
        plt.imshow(x_adv[0])
        plt.show(block=False)

    attack.max_iter = iter_step

```
But, when it is clearly seen that in the picture "disagreement" is still recognized as "agreement",

![TF](https://user-images.githubusercontent.com/98591432/152143737-301682d5-f996-4fca-a604-e451c9523280.png)



 an error occurs during a further attack:

```
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_2340/1440880561.py in <module>
      3 x_adv = np.array([image_init], dtype=np.float32)
      4 for i in range(32):
----> 5     x_adv = attack.generate(x=np.array([image_target], dtype=np.float32), x_adv_init=x_adv, y=to_categorical([1], 3))
      6 
      7     #clear_output()

~\anaconda3\lib\site-packages\art\attacks\evasion\hop_skip_jump.py in generate(self, x, y, **kwargs)
    190 
    191             if self.targeted:
--> 192                 x_adv[ind] = self._perturb(
    193                     x=val,
    194                     y=y[ind],

~\anaconda3\lib\site-packages\art\attacks\evasion\hop_skip_jump.py in _perturb(self, x, y, y_p, init_pred, adv_init, mask, clip_min, clip_max)
    257 
    258         # If an initial adversarial example found, then go with HopSkipJump attack
--> 259         x_adv = self._attack(initial_sample[0], x, initial_sample[1], mask, clip_min, clip_max)
    260 
    261         return x_adv

~\anaconda3\lib\site-packages\art\attacks\evasion\hop_skip_jump.py in _attack(self, initial_sample, original_sample, target, mask, clip_min, clip_max)
    436                 epsilon /= 2.0
    437                 potential_sample = current_sample + epsilon * update
--> 438                 success = self._adversarial_satisfactory(
    439                     samples=potential_sample[None],
    440                     target=target,

~\anaconda3\lib\site-packages\art\attacks\evasion\hop_skip_jump.py in _adversarial_satisfactory(self, samples, target, clip_min, clip_max)
    637         """
    638         samples = np.clip(samples, clip_min, clip_max)
--> 639         preds = np.argmax(self.estimator.predict(samples, batch_size=self.batch_size), axis=1)
    640 
    641         if self.targeted:

~\anaconda3\lib\site-packages\art\estimators\classification\classifier.py in replacement_function(self, *args, **kwargs)
     69                 if len(args) > 0:
     70                     args = tuple(lst)
---> 71                 return fdict[func_name](self, *args, **kwargs)
     72 
     73             replacement_function.__doc__ = fdict[func_name].__doc__

~\anaconda3\lib\site-packages\art\estimators\classification\blackbox.py in predict(self, x, batch_size, **kwargs)
    130                 min((batch_index + 1) * batch_size, x_preprocessed.shape[0]),
    131             )
--> 132             predictions[begin:end] = self.predict_fn(x_preprocessed[begin:end])
    133 
    134         # Apply postprocessing

~\AppData\Local\Temp/ipykernel_2340/2349775553.py in predict(x)
     16         # read text
     17         file = open("out.txt","r+")
---> 18         test = file.read()
     19         out_string = test.strip()
     20 

~\anaconda3\lib\encodings\cp1251.py in decode(self, input, final)
     21 class IncrementalDecoder(codecs.IncrementalDecoder):
     22     def decode(self, input, final=False):
---> 23         return codecs.charmap_decode(input,self.errors,decoding_table)[0]
     24 
     25 class StreamWriter(Codec,codecs.StreamWriter):
UnicodeDecodeError: 'charmap' codec can't decode byte 0x98 in position 2: character maps to <undefined>
```

**UnicodeDecodeError: 'charmap' codec can't decode byte 0x98 in position 2: character maps to <undefined> **

**Reply by beat-buesser** [reply]
@HexHexHex16 Are you able to extract the file `out.txt` created by Tesseract and show its content? It should contain the current prediction by Tesseract but it seems that it now contains a character that Python doesn't expect. This could indicate that the prediction has changed but that we need to adapt the character map used to read the file.

**Reply by HexHexHex16** [reply]
Yes, here is the contents of the file out.txt ( `‘dissent` )

**Reply by beat-buesser** [reply]
I think to continue iterating the attack to reduce the perturbation further we would have to set the correct endoding in the call to `open` in cell 3. The content of out.txt also shows that the perturbation of the last iteration isn't able to fool Tesseract to return "assent" anymore, but " 'dissent" which is closer to the "dissent" shown in the image, therefore less adversarial.

**Reply by HexHexHex16** [reply]
How exactly can I eliminate this character in ` ['dissent]` by modifying the open() method in cell 3?

**Reply by beat-buesser** [reply]
Maybe a different encoding in the call to `open` might help.
